pkg/operator/controller/controller.go

```diff
@@ -283,15 +300,25 @@ func (r *reconciler) ensureDNSDeleted(dns *operatorv1.DNS) error {
 // ensureDNSNamespace ensures all the necessary scaffolding exists for
 // dns generally, including a namespace and all RBAC setup.
 func (r *reconciler) ensureDNSNamespace() error {
-	ns := manifests.DNSNamespace()
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
+	existingNamespace := corev1.Namespace{}
+	desiredNamespace := manifests.DNSNamespace()
+	if err := r.client.Get(context.TODO(), DefaultDNSOperandNamespaceName(), &existingNamespace); err != nil {
 		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to get dns namespace %q: %v", ns.Name, err)
+			return fmt.Errorf("failed to get dns namespace %q: %w", desiredNamespace.Name, err)
 		}
-		if err := r.client.Create(context.TODO(), ns); err != nil {
-			return fmt.Errorf("failed to create dns namespace %s: %v", ns.Name, err)
+		if err := r.client.Create(context.TODO(), desiredNamespace); err != nil {
+			return fmt.Errorf("failed to create dns namespace %s: %w", desiredNamespace.Name, err)
+		}
+		logrus.Infof("created dns namespace: %s", desiredNamespace.Name)
+	} else {
+		// Make sure existing DNS namespace has all the labels from the Namespace manifest
+		changed := namespaceLabelsChanged(&existingNamespace, *desiredNamespace)
```

Comment:
So looks like this `namespaceLabelsChanged` function is mutating `existingNamespace`, which I feel is a bit unexpected, as it reads as a boolean-style function. If we look at the pattern of other object `____Changed` functions, they don't mutate the existing object: [canaryNamespaceChanged](https://github.com/openshift/cluster-ingress-operator/blob/542124811a01039852d8b0846da91d33665037f8/pkg/operator/controller/canary/namespace.go#L82) or [dnsRecordChanged](https://github.com/openshift/cluster-ingress-operator/blob/ce263ce12526a398ae36188a1bb59c40e847787d/pkg/operator/controller/ingress/dns.go#L187).

The pattern I see in our code is `ensureObject` -> `updateObject` -> `objectChanged`. A very similar use case may be [ensureCanaryNamespace](https://github.com/openshift/cluster-ingress-operator/blob/542124811a01039852d8b0846da91d33665037f8/pkg/operator/controller/canary/namespace.go#L18) or [ensureRouterNamespace](https://github.com/openshift/cluster-ingress-operator/blob/ce263ce12526a398ae36188a1bb59c40e847787d/pkg/operator/controller/ingress/namespace.go#L20) I'd consider adapting the patterns there to be consistent.

Reply:
I'm going to decline, hoping this is not a sticking point.  cluster-dns-operator is inconsistent with cluster-ingress-operator in **many** ways, especially in the `Reconcile`, and I don't want to start refactoring as a part of this bug fix.  The `namespaceLabelsChanged` code  was modeled after `daemonsetConfigChanged` in controller_dns_daemonset.go.

Reply:
Actually I thought I was in cluster-ingress-operator, so I agree with your consistency point. cluster-dns-operator is more inconsistent. Regarding the whole ensure -> Update -> changed functional pattern, that's not a problem.

But to your point, `daemonsetConfigChanged` doesn't mutate the `current` value. Consistency aside, I think `namespaceLabelChanged` is violating expectation here. Would you consider matching the pattern of `daemonsetConfigChanged` like, it should be a simple update:
```suggestion
		changed, updated := namespaceLabelsChanged(*existingNamespace, *desiredNamespace)
```